FWCore/Catalog/src/FileLocator.cc

```diff
@@ -293,22 +285,27 @@ namespace edm {
 
     std::string protName = found_protocol->second.get("protocol", kEmptyString);
     m_protocols.push_back(protName);
-    m_prefix = found_protocol->second.get("prefix", kEmptyString);
-    if (m_prefix == kEmptyString) {
-      //get rules
-      if (found_protocol->second.find("rules") == found_protocol->second.not_found()) {
-        cms::Exception ex("FileCatalog");
-        ex << "protocol must contain either a prefix or rules, "
-           << "neither found for protocol \"" << aCatalog.protocol << "\" for the storage site \""
-           << aCatalog.storageSite << "\" and volume \"" << aCatalog.volume
-           << "\" in storage.json. Check site-local-config.xml <data-access> and storage.json";
-        ex.addContext("edm::FileLocator:init()");
-        throw ex;
-      }
-      const pt::ptree& rules = found_protocol->second.find("rules")->second;
+
+    //store all prefixes and rules to m_directRules. We need to do this so that "applyRules" can find the rule in case chaining is used
+    //loop over protocols
+    for (pt::ptree::value_type const& protocol : protocols) {
+      std::string protName = protocol.second.get("protocol", kEmptyString);
       //loop over rules
-      for (pt::ptree::value_type const& storageRule : rules) {
-        parseRule(storageRule, protName, m_directRules);
+      std::string prefixTmp = protocol.second.get("prefix", kEmptyString);
+      if (prefixTmp == kEmptyString) {
+        const pt::ptree& rules = protocol.second.find("rules")->second;
+        for (pt::ptree::value_type const& storageRule : rules) {
+          parseRule(storageRule, protName, m_directRules);
+        }
+      }
+      //now convert prefix to a rule and save it
+      else {
+        Rule rule;
+        rule.pathMatch.assign("(/.*)");  //require lfn start with "/"
+        rule.destinationMatch.assign(".*");
+        rule.result = prefixTmp + "$1";
```

Comment:
Given the check on lines 71-72 already requires the LFN to start with `/store/`, this could be
```suggestion
        rule.pathMatch.assign("("+kLFNPrefix+".*)");
        rule.destinationMatch.assign(".*");
        rule.result = prefixTmp + "$1";
```
that could be more clear wrt. the check on lines 71-72.


Just a note fore future reference that the earlier code (lines 73-80) had a logic to add `/` between prefix and LFN in case the `LFN` would not have started with `/`, but that was actually unnecessary given the check on lines 71-72.

Reply:
I think it is better to reserve the logic of prefix, which is simply to be placed in front of lfn and let the enforcement of what is a valid lfn separately somewhere when needed. Therefore I will write:
rule.pathMatch.assign("(.*)");
rule.destinationMatch.assign(".*");
rule.result = prefixTmp + "/$1";

Reply:
> Therefore I will write:
> 
> ```
> rule.pathMatch.assign("(.*)");
> rule.destinationMatch.assign(".\*");
> rule.result = prefixTmp + "/$1";
> ```

This would then always add the `/` even when the LFN starts with a slash (i.e. in practice always). Is that intentional? How about
```cpp
rule.pathMatch.assign("/?(.*)");
rule.destinationMatch.assign(".*");
rule.result = prefixTmp + "/$1";
```
instead?

Reply:
Sorry, my bad! If I understand correctly from Stephan, we do not want to add ``/`` between prefix and lfn if lfn already starts with ``/``. That is why have had redundant lines 73-80 earlier. Your suggestion is ideal!